gist.go

```diff
@@ -0,0 +1,86 @@
+package model
+
+// Typed constants
+type ChannelType string
+
+// CamelCase Constant Names
+const (
+	ChannelTypeOpen    ChannelType = "O"
+	ChannelTypePrivate ChannelType = "P"
+)
+
+type Channel struct {
+	// Member variables are not exported to ensure immutability.
+	id       string `json:"id"`
+	name     string `json:"name"`
+	createAt int64  `json:"create_at"`
+	// etc...
+
+	// Question: what do we do when there's a semantically meaningful
+	// "null" value for a member variable?
+	schemeID *string `json:"scheme_id"`
```

Comment:
I think we start by challenging whether there are any legitimate semantics for nil primitives. I recognize database record sets and such use them, but I don't think we should let nullable (nilable) types propagate through the code.  Either a new type is needed 
```go
type ID struct {
    id string
   empty bool
}   
```
or we deal with Go's zero values.

Reply:
Yeah, that's a good point. I'll take a look and see just how many of these cases exist at the moment and we can take it from there.

Reply:
Here's a list of the fields I found (in the v5.23 branch)

```
BundleInfo.Manifest
Channel.SchemeId
Channel.GroupConstrained
ChannelsWithCount.Channels
ChannelForExport.SchemeName
DirectChannelForExport.Members
ChannelModeration.Roles
ChannelModeratedRoles.Guests
ChannelModeratedRoles.Members
ChannelSearchOpts.Page
ChannelSearchOpts.PerPage
ChannelData.Channel
ChannelData.ChannelMember
ChannelMemberHistory.LeaveTime
ChannelMemberHistoryResult.LeaveTime
ChannelSearch.Page
ChannelSearch.PerPage
Group.MemberCount
GroupWithSchemeAdmin.SchemeAdmin
LdapGroupSearchOpts.IsLinked
LdapGroupSearchOpts.IsConfigured
GroupSearchOpts.PageOpts
InitialLoad.User
PostAction.PostActionIntergration
PostActionCookie.PostActionIntegration
PostActionIntegrationResponse.Update
BaseMarketplacePlugin.Manifest
MarketplacePlugin.BaseMarketplacePlugin
OutgoingWebhookResponse.Text
.... and around here I basically ran out of steam with this.
```

There are also instances of slices of Pointers (especially when nesting a list of one model struct inside another model struct). And then there's the contents of `config.go` and `licence.go`. Also `manifest.go` seems to be quite a deep rabit-hole of pointers, and `message_export.go`.

Worryingly, I get the sense that the older models don't use pointers so much, whereas the newer ones do use them more often. I have a really horrible feeling about who might be responsible for starting that trend... oops.

Reply:
So, in conclusion, I think we need to pull out a few of these examples, and figure out how we'd treat them in the new model approach. Then maybe we can start to generalise from there. @wiggin77 or anyone else - any further thoughts on how to progress this particular issue?

Reply:
For any `*Patch` objects, and in the case of string properties, we definitely need to distinguish `nil` from the zero-value, since we want to distinguish not changing the current value from setting it to an empty string. Either way, I agree that this should be rare, and likely not needed in most cases.

Reply:
(I see this exact thing is discussed later on in the PR!)